### PR TITLE
Processing data models

### PR DESCRIPTION
**Pull request by dev89s**
1. Created model classes for all entities
2. Created associations between models
3. Added some entries on the rails console to check if the model classes work properly
4. Added custom methods to model classes
5. Added some entities and tested the methods in the rails console

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,3 +1,11 @@
 class User < ApplicationRecord
-  has_many posts
+  has_many :posts, foreign_key: :author_id
+  has_many :likes, foreign_key: :author_id
+  has_many :comments, foreign_key: :author_id
+  has_many :liked_posts, through: :likes, source: :post
+  has_many :commented_posts, through: :comments, source: :post
```
Comment (mckent05) [inline app/models/user.rb:6]: **OPTIONAL** - I couldn;t find a join table in your schema file, so I do not think this code is needed. Please remove his to help make your code easier to read and maintain 👍 👍

File: app/models/post.rb [hunk 2]
```diff
@@ -1,5 +1,15 @@
 class Post < ApplicationRecord
-  belongs_to :users, counter_cache: :posts_counter
-  has_many :likes
-  has_many :comments
+  belongs_to :author, counter_cache: :posts_counter, class_name: 'User'
+  has_many :likes, foreign_key: :post_id
+  has_many :comments, foreign_key: :post_id
+  # has_many :users, through: :likes
+  # has_many :users, through: :comments
```
Comment (mckent05) [inline app/models/post.rb]: - Please remove all commented code from your codebase to help make your code easier to read and maintain 👍 👍

File: app/models/post.rb [hunk 3]
```diff
@@ -1,5 +1,15 @@
...
+
+  def post_counter_update
+    author.update(posts_counter: author.posts.count)
```
Comment (mckent05) [inline app/models/post.rb:8]: -I suggest that you call the post_counter_update when the data is saved into the database, please use the `after_save` method to fix this 🙏  i.e `after_save :post_counter_update`

File: app/models/like.rb [hunk 4]
```diff
@@ -1,3 +1,8 @@
 class Like < ApplicationRecord
-  belongs_to :posts, counter_cache: :likes_counter
+  belongs_to :post, counter_cache: :likes_counter
+  belongs_to :author, class_name: 'User'
+
+  def like_counter_update
+    post.update(likes_counter: post.likes.count)
+  end
```
Comment (mckent05) [inline app/models/like.rb:8]: -I suggest that you call the likes_counter_update when the data is saved into the database, please use the `after_save` method to fix this 🙏  i.e `after_save :like_counter_update`

File: app/models/comment.rb [hunk 5]
```diff
@@ -1,3 +1,8 @@
 class Comment < ApplicationRecord
-  belongs_to :posts, counter_cache: :comments_counter
+  belongs_to :post, counter_cache: :comments_counter
+  belongs_to :author, class_name: 'User'
+
+  def update_comments_counter
+    post.update(comments_counter: post.comments.count)
+  end
```
Comment (mckent05) [inline app/models/comment.rb:8]: -I suggest that you call the update_comments_counter when the data is saved into the database, please use the `after_save` method to fix this 🙏  i.e `after_save :update_comments_counter`

**Review by mckent05 — CHANGES_REQUESTED** [review]
Hi @dev89s, 

# STATUS: CHANGES REQUIRED ♻️ ♻️ 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
almost there

## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👍 👍
- Good job setting up the associations between your models 👍 👍  


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/post.rb (hunk 2 above)
Comment (Abenezer-Tilahun) [reply app/models/post.rb]: Fixed.

File: app/models/post.rb (hunk 3 above)
Comment (Abenezer-Tilahun) [reply app/models/post.rb:8]: Fixed.

File: app/models/like.rb (hunk 4 above)
Comment (Abenezer-Tilahun) [reply app/models/like.rb:8]: Fixed.

File: app/models/comment.rb (hunk 5 above)
Comment (Abenezer-Tilahun) [reply app/models/comment.rb:8]: Fixed.

**Review by Abenezer-Tilahun — APPROVED** [review]
Hi @dev89s 
 
Your project is complete  👍 💯   There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 👏🏻

✔️ No linter errors.
✔️ You used correct Git-Flow.
✔️ Professional README  
✔️ Good Commit Messages
✔️ Good job for Implemented Suggested changes.

## Optional suggestions 
 
_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

**Pull request merged**